python/libraries/bimvee/geometry.py

```diff
@@ -147,6 +147,35 @@ def combineTwoQuaternions(q1, q2):
     qOut[1:4] = vOut
     return qOut
 
+def quaternionConjugate(q):
+    return np.array([q[0], -q[1], -q[2], -q[3]])
+
+def quaternionInverse(q):
+    return quaternionConjugate(q) / np.sum(q ** 2)
+
+def quaternionProduct(q1, q2):
+    return np.array([ 
+    q1[0] * q2[0] - q1[1] * q2[1] - q1[2] * q2[2] - q1[3] * q2[3],
+    q1[0] * q2[1] + q1[1] * q2[0] + q1[2] * q2[3] - q1[3] * q2[2],
+    q1[0] * q2[2] - q1[1] * q2[3] + q1[2] * q2[0] + q1[3] * q2[1],
+    q1[0] * q2[3] + q1[1] * q2[2] - q1[2] * q2[1] + q1[3] * q2[0] ])
+
+def angleBetweenTwoQuaternions(q1, q2):
+    # returns minimal angle in radians between two unit quaternions, following:
+    # https://www.researchgate.net/post/How_do_I_calculate_the_smallest_angle_between_two_quaternions    
+    qP = quaternionProduct(q1, quaternionInverse(q2))
+    normV = np.linalg.norm(qP[1:])
+    return 2 * np.arcsin(normV)
+
+def averageOfQuaternions(allQ):
+    # follows: http://wiki.unity3d.com/index.php/Averaging_Quaternions_and_Vectors?_ga=2.89251931.1981359277.1588961760-2124566047.1586341788
+    qAvg = allQ[0, :]
+    for idx, nextQ in enumerate(allQ[1:, :]):
+        weight = 1.0 / (idx+2)
+        qAvg = slerp(qAvg, nextQ, weight)
```

Comment:
Is this correct? Not sure if average of a set of quaternions has a special meaning but it seems like the avg at each step is something like:
avg = (q0+q1)/2
avg = ((q0+q1)/2 + q3)/3
avg = (((q0+q1)/2 + q3)/3 +q4)/4
...

Reply:
That's right - it's just applying slerp between all quaternions, one pair at a time and weighting accordingly.

Reply:
Shouldn't the avg at the nth step be something like (q0 + q1 + ... qn) / n ?

Reply:
No: 

`q1 = np.array([1, 0, 0, 0])
q2 = np.array([-1, 0, 0, 0])
allQ = np.concatenate([q1[np.newaxis, :], q2[np.newaxis, :]], axis = 0)
thisMethod = bimvee.geometry.averageOfQuaternions(allQ)
naiveMethod = np.mean(allQ, axis=0)
`

Reply:
Ah now I understand your previous comment - the problem is when the components wrap around e.g. from 1-> -1, then naive averaging doesn't work.

Reply:
... good catch - corrected this function.